osbuild/util/pathfd.py

```diff
@@ -0,0 +1,716 @@
+
+        self._fd = fd
+
+    @classmethod
+    def from_path(cls, path, *, follow_symlink = False):
+        """Create Path Descriptor from File System Path
+
+        This constructor creates a new path-descriptor from a standard
+        file-system path provided as a string. It simply opens the path as
+        `O_PATH` file-descriptor and creates a wrapping path-descriptor around
+        it. Path resolution happens at the time of this function call. All
+        further operations on the new path-descriptor will happen relative to
+        the desciptor itself. The contents of `path` are not retained and will
+        not be accessible later on.
+
+        Parameters
+        ----------
+        path : String
+            A file-system path in string representation. If it is a relative
+            path, it is interpreted relative to the current working directory
+            at the time of this function call.
+        follow_symlink : Boolean, optional
+            Whether to resolve the final part of the path in case it is a
+            symlink. The default behavior is to not resolve it. That is, if the
+            given path points to a symlink, the path-descriptor will also point
+            to the symlink, rather than pointing to the destination of that
+            symlink.
+            This does not affect symlinks somewhere else in the path, other
+            than the last element.
+
+        Returns
+        -------
+        PathFd
+            A new path-desciptor for the specified path is returned.
+
+        Raises
+        ------
+        OSError
+            Any operating system errors involved in resolving file-system paths
+            can be raised by this function. See `os.open()` in the python
+            standard library for a discussion.
+        """
+        flags = _os.O_CLOEXEC | _os.O_PATH
+        if not follow_symlink:
+            flags |= _os.O_NOFOLLOW
+
+        return PathFd(_os.open(path, flags))
+
+    def close(self):
+        """Close the Path Descriptor
+
+        This closes the path-descriptor and releases all pinned resources. Once
+        this returns, any further operation on the path-descriptor will raise
+        an exception, unless it is explictily supported on closed descriptors.
+
+        This function can be safely called on closed descriptors, in which case
+        it is a no-op.
+        """
+
+        if self._fd is not None:
+            _os.close(self._fd)
+            self._fd = None
+
+    def __del__(self):
+        # Make sure to release internal resources when variables of this type
+        # are cleaned up by the GC. In case they were already explicitly closed
+        # this will be a no-op.
+        self.close()
+
+    def __enter__(self):
+        # Trivial context-manager ala `contextlib.closing()`.
+        return self
+
+    def __exit__(self, *args):
+        # Trivial context-manager ala `contextlib.closing()`.
+        self.close()
+
+    def __int__(self):
```

Comment:
I read through the cpython sources because I wanted the `dirfd=` arguments to automatically do the int conversion. It never worked, but I kinda kept the `__int__()` overloading. I was not aware that `fileno()` is the canonical way to do that. I just checked, it is even hard-coded in `int PyObject_AsFileDescriptor(PyObject *p)`. Maybe this will do the conversion automatically. I will check! And I will definitely change it, as that sounds like the better option.